### PR TITLE
Add tolerances for float comparison in VWActionScoresLearnerTest tests

### PR DESCRIPTION
**Pull request by jackgerrits**
In these tests the ActionScores are compared using direct equality, this change introduces a tolerance to the float comparisons to deal with rounding precision errors. This resolves test issues seen in #1706 

The tolerance is taken from the minor allowable tolerance in RunTests.

### CODE REVIEW

**Comment by jackgerrits** [comment]
Comparing floats exactly is rarely reliable due to rounding errors while being operated on, this just makes the test resilient to this.

**Comment by JohnLangford** [comment]
Comparing float values which are equal as reals but computed via a different sequence of operations is not reliable due to rounding.   Where is this happening?

The worry here: my experience is that float values in tests which are not _precisely_ equal end up creating cascading differences of arbitrary scale due to tie-breaking situations.  So, I don't want to paper over an issue that should be nailed down now.

File: java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java [hunk 1]
```diff
@@ -4,16 +4,65 @@
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
 import vowpalWabbit.VWTestHelper;
+import vowpalWabbit.responses.ActionScore;
 import vowpalWabbit.responses.ActionScores;
 
 import java.io.IOException;
 
 import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertTrue;
```
Comment (jmorra) [inline java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java]:
```suggestion
import static org.junit.Assert.assertEquals;
```

File: java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java [hunk 2]
```diff
@@ -4,16 +4,65 @@
...
 
 /**
  * Created by jmorra on 10/2/15.
  */
 public class VWActionScoresLearnerTest extends VWTestHelper {
+    private final float TOLERANCE = 0.001f;
```
Comment (jmorra) [inline java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java]: None of this is necessary as it is already in `Assert`.

File: java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java [hunk 3]
```diff
@@ -61,7 +110,7 @@ public void testCSOAA() throws IOException {
                         actionScore(1, 1.0227613f)
                 )
         };
-        assertArrayEquals(expected, pred);
+        assertTrue(equalWithTolerance(expected, pred, TOLERANCE));
```
Comment (jmorra) [inline java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java]:
```suggestion
        assertEquals(equalWithTolerance(expected, pred, TOLERANCE));
```

File: java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java [hunk 4]
```diff
@@ -110,7 +159,7 @@ private void testCBADF(boolean withRank) throws IOException {
             )
         };
         vw.close();
-        assertArrayEquals(expectedTrainPreds, trainPreds);
+        assertTrue(equalWithTolerance(expectedTrainPreds, trainPreds, TOLERANCE));
```
Comment (jmorra) [inline java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java]:
```suggestion
        assertEquals(equalWithTolerance(expectedTrainPreds, trainPreds, TOLERANCE));
```

File: java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java [hunk 5]
```diff
@@ -123,6 +172,6 @@ private void testCBADF(boolean withRank) throws IOException {
         };
 
         vw.close();
-        assertArrayEquals(expectedTestPreds, testPreds);
+        assertTrue(equalWithTolerance(expectedTestPreds, testPreds, TOLERANCE));
```
Comment (jmorra) [inline java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java]:
```suggestion
        assertEquals(equalWithTolerance(expectedTestPreds, testPreds, TOLERANCE));
```

**Comment by jackgerrits** [comment]
@jmorra the assertEquals(Object, Object) overload is being used as it is not directly comparing a float. That's why I had to actually iterate over the objects to compare the values.

File: java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java (hunk 2 above)
Comment (jackgerrits) [reply java/src/test/java/vowpalWabbit/learner/VWActionScoresLearnerTest.java]: See comment below about using the `Object` overload meaning tolerance cannot be used

**Comment by jackgerrits** [comment]
@JohnLangford I have tried to reproduce this difference in values for quite some time but cannot nail down what is causing it. As far as I can tell the prediction values are identical in the cpp layer. I created an app doing the equivalent operations in cpp only and the values for the prediction before and after the options change are identical (on Windows, so it may be a Linux only difference). The only change I made to the Java binding layer was an exception being translated at the binding layer.

**Comment by jmorra** [comment]
It would appear that a newer version of JUnit might have this functionality built in.

https://stackoverflow.com/questions/37627340/junit-assert-double-arrays

You would have to just pull out the values and you can do that generically.

**Comment by jackgerrits** [comment]
Ah I see, so flatten and compare? I'll update

**Comment by jackgerrits** [comment]
Closing in favor of #1709, this was actually a bug with the implementation of `exp` that was being linked. It seems as if we are okay not using tolerances in these tests for now.

**Pull request closed without merging**